samples/matter/common/src/diagnostic/diagnostic_logs_crash.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#pragma once
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+namespace Nrf
+{
+struct CrashDescription {
+	z_arch_esf_t Esf = {};
+	unsigned int Reason = 0;
+	const char *ThreadName = nullptr;
+	uint32_t *ThreadInt = 0;
+	unsigned int Source = 0;
+};
+
+#ifdef CONFIG_NCS_SAMPLE_MATTER_DIAGNOSTIC_LOGS_CRASH_LOGS
+/* Assuming 2 Bytes for prefix and 2 bytes for CRC */
+static_assert(DT_REG_SIZE(DT_NODELABEL(crash_retention)) - 4 > sizeof(CrashDescription));
+#endif
+
+/**
+ * @brief Class for creating log from the CrashData data.
+ *
+ * CrashData data can be obtained from the fatal error ISR and converted to human-readable
+   format.
+ *
+ * This class takes care on the conversion from ESF data to Log and
+ *
+ * There is several steps of converting the esf data, and some of them can be disabled using kconfig
+   to reduce the final log size:
+
+	* BasicDump converts the basic registers of the device.
+	  Cannot be disabled.
+	* FpuDump converts the floating point unit registers of the device.
+	  Can be disabled using CONFIG_FPU and CONFIG_FPU_SHARING
+	* ExtraDump converts an additional extra exception info registers of the device.
+	  Can be disabled using CONFIG_EXTRA_EXCEPTION_INFO.
+	* ReasonDump converts the reason of the fault.
+	  Cannot be disabled.
+	* MultithreadDump converts the information about the the thread from which the error occurred.
+	  Can be disabled using CONFIG_MULTITHREADING.
+ */
+class CrashData {
+public:
+	constexpr static size_t MaxSingleDumpSize = 100;
+
+	/**
+	 * @brief Convert CrashDat to output string buffer that will contain a crash message.
+	 *
+	 * This method converts the crash data received from the ISR to human-readable
+	   format that can be displayed as log.
+	 *
+	 * @param outBuffer output string buffer
+	 * @param bufferSize size of the output string buffer.
+	 * @param end Assign that this is end of the whole CrashData stored.
+	 * @return size_t as the actual length of the converted data.
+	 */
+	size_t ProcessConversionToLog(char *outBuffer, size_t bufferSize, bool &end);
+
+	/**
+	 * @brief Get a pointer to the crash description structure.
+	 *
+	 * This method can be used to fill CrashDescription data.
+	 *
+	 * @return CrashDescription* a pointer to crash description object that contains ESF and reason data.
+	 */
+	CrashDescription *GetDescription() { return &mDescription; }
+
+	/**
+	 * @brief Calculate a size of the actual crash logs.
+	 *
+	 * The actual size depends on the crash type and collected crash data.
+	 *
+	 * @return size_t size of the actual crash logs.
+	 */
+	size_t CalculateSize();
+
+private:
+	enum FaultSource {
+		HardFault = 3,
```

Comment:
Are these values taken from some other reference? If so, maybe it's good to add a comment with reference.

Reply:
As we discussed F2F, it's kind of hard to find the reference. I put the generic comment explaining that values are based on vector table.